test/test-parse-statement.cpp

```diff
@@ -578,6 +578,44 @@ TEST(test_parse, else_without_if) {
   }
 }
 
+TEST(test_parse, else_unexpected_semicolon) {
+  {
+    spy_visitor v;
+    padded_string code(u8"if (cond) { body; } else; { body; }"_sv);
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.visits, ElementsAre("visit_variable_use",        // cond
+                                      "visit_enter_block_scope",   // (if)
+                                      "visit_variable_use",        // body
+                                      "visit_exit_block_scope"));  // (if)
+
+    EXPECT_THAT(v.errors,
+                ElementsAre(ERROR_TYPE_FIELD(
+                    error_unexpected_semicolon_after_else, semicolon,
+                    offsets_matcher(&code, strlen(u8"if (cond) { body; } else"),
+                                    u8";"))));
+  }
+}
+
+TEST(test_parse, else_has_empty_body) {
+  {
+    spy_visitor v;
+    padded_string code(u8"if (cond) { body; } else;\n{ }"_sv);
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_THAT(v.visits, ElementsAre("visit_variable_use",        // cond
+                                      "visit_enter_block_scope",   // (if)
+                                      "visit_variable_use",        // body
+                                      "visit_exit_block_scope"));  // (if)
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(
+            error_else_has_empty_body, where,
+            offsets_matcher(&code, strlen(u8"if (cond) { body; } else {"),
+                            u8"}"))));
```

Comment:
Blocking: This strlen string doesn't match the code. The `u8"}"` looks wrong too. Did you mean the following instead?
```suggestion
            offsets_matcher(&code, strlen(u8"if (cond) { body; } else;\n"),
                            u8"{"))));
```

If so, the error looks misplaced. It should point at the `else` token or the `;`, not at the following unrelated `{`.

Reply:
Right now it detects `else;{}` (semicolon after else) and `else;\n{}` (empty body), the following in [parse.h:2534](https://github.com/quick-lint/quick-lint-js/pull/290/files#diff-660547dad34c381d590ccfcc6343d98eb7c4b235bec2dc324fecd1cb62f71662R2523) should change to only trigger **_else has empty body_** when it is not preceded by a newline:

``` diff 
-   if (token_after_else.type == token_type::semicolon &&
-          this->peek().type == token_type::left_curly) {
+   if (token_after_else.type == token_type::semicolon &&
+          this->peek().type == token_type::left_curly &&
+          !this->peek().has_leading_newline) {

-     if (this->peek().has_leading_newline) {
+     if (body is empty) {
          this->error_reporter_->report(
              error_else_has_empty_body{.where = this->peek().span()});

       } else {
          this->error_reporter_->report(error_unexpected_semicolon_after_else{
              .semicolon = token_after_else.span()});
        }
    }
```

Should I omit the else has empty body warning altogether, or should I detect the empty body?
I do not know how to do it without making use of a transaction, advancing the peek and then doing a rollback afterwards.

Reply:
> Should I omit the else has empty body warning altogether

Yeah, let's only emit error_unexpected_semicolon_after_else in this PR.

> I do not know how to do it without making use of a transaction, advancing the peek and then doing a rollback afterwards.

Would something like this work? Or am I missing an edge case?
```c++
      if (token_after_else.type == token_type::semicolon &&
          this->peek().type == token_type::left_curly &&
          !this->peek().has_leading_newline) {
        this->error_reporter_->report(error_unexpected_semicolon_after_else{
          .semicolon = token_after_else.span(),
        });
      }
```